### ISSUE TITLE
[Not an issue] About FW binaries

### ISSUE THREAD

**Issue by Lar-Sen** [bleem313/amiloop-files#1]
Hello

Got an AmiiboLink (hw ver: marked "LC_KGF_NT_V2.5"
Tried to reflash FW image with your v3_flash.bin at first.
Device only bootup if I tickle some capacitors (otherwise LED stays on and nothing happens)

..they surely did something to prevent crossflashing

Tried to reflash FW with your v4_230608_flash.bin.
Device hangs (hw crash, visible by probing device with blackmagic probe)

Tried to mix app binary, erasing @0x26000 to the end, then pushing @0x26000 the amiibolink(release)_v1.2.1.33_202304191123.bin found somewhere else : It works.

But I still haven't managed to convert my AmiiboLink to Amiloop.

Question: How did you dump the v4_230608_flash.bin ?  Flash is locked... We have only access to "Nordic nRF52 Access Port" which of course doesn't allow memory dump, only mass erasing.
Do you know where I can find a full dump of AmiiboLink ?  I would like to experiment further.

Btw I noticed only v4 has DFU support. v3 is not FOTA compatible.

**Comment by likaci** [comment]
I came here from reddit, and [in this post](https://www.reddit.com/r/Amiibomb/comments/11rmwmi/comment/jq9aoe7/) author mentions using an ESP32 to glitch nrf52 and dump the data https://github.com/atc1441/ESP32_nRF52_SWD

I have an Amiloop and an AmiiboLink, they both can flash v3 v4 and work well.
I also tried to go back to the AmiiboLink version, but failed, maybe I can try your method.

**Comment by bleem313** [comment]
@likaci is correct. This vulnerability was fixed in newer revisions of the chip, so if you attempt it your mileage may vary. I think the Amiloop purposefully uses the older revision as they would be cheaper.

I don't have an AmiiboLink to work with, so I can't test converting it, but I believe others have. You may want to check your connection and make sure it is stable. You are correct that the v3 firmware doesn't have OTA enabled. That was one of the main reasons I started this project, as buying the v3 was much cheaper than a v4.

**Comment by likaci** [comment]
I tried @Lar-Sen method, and it does work.
```
flash write_image erase v3_flash.bin
flash erase_sector 0 38 127
flash write_image erase amiibolink(release)_v1.2.1.33_202304191123.bin 0x26000
```
After that it can be recognized as AmiiboLink by NFC TAG Pro App, and the flash will also be protected.

AmiiboLink dfu file can be found here, https://github.com/strengthen/iNFCAmiibo/tree/main/Firmware/AmiiboLink
and

**Comment by Lar-Sen** [comment]
@likaci , thank you for the link 
I was aware of the glitching possibility up to AAE1 revision of the chip, which first came to light thanks to [this guy](https://limitedresults.com/2020/06/nrf52-debug-resurrection-approtect-bypass/) 's work. Anyway going cheaper like atc441 did is brilliant.

It seems low prices observed on Ali for these devices could be related to earlier revisions of the nRF52832 without the errata. All PCB pictures found mention AAE0 or AAE1 (strange one, not even mentioned in official reference docs... Are they counterfeit?) which are all rev.2 chips. Power glitch was fixed in v3.

There are in fact some component differences between AmiiboLink and Amiloop : I noticed 2 missing capacitors in the "V2.5" PCB. This could explain why Amiloop v3 firmware is hardly bootable, and v4 fails on it.

In short you can flash AmiiboLink fw on Amiloop but not the other way around.

In the meantime I tried to supply AmiLoop bootloader to my AmiiboLink PCB (adding the 2 missing offsets at the very end of the MBR section): It fails to boot the app, and stays stuck in DFU rescue mode. No surprise here, as I was unable to find a way to update app's signature to validate boot. Anyway CRC checks are easily corrected via nrfutil.

As a bonus I provide an easy patch to app binary, to disable APPROTECT anytime. Can be useful to debug, or quickly dump perfectly fine Amiibo's stuck in flash.
At offset 0x7DC (ROM: 0x267DC) overwrite with these bytes:
`5A 28 0E D0 07 4B 01 20 18 60 07 48 01 68 00 29 FC D0 5A 21 C2 F8 08 12 00 21 19 60 01 68 00 29 FC D0 70 47`
(this is for amiibolink v1.2.1.33)

**Comment by likaci** [comment]
@Lar-Sen thank you for the bonus, it work like a charm :)
After searching [LC_KGF_NT_V2.5 PCB photo](https://imgur.com/gallery/72CGuWt), I found that my AmiiboLink is the same layout, but it can run the v3 v4 Amiloop bin. This is mine, is it the same as yours?
<img src="https://github.com/bleem313/amiloop-files/assets/3407980/1189ca48-26c3-4277-a9cf-d60a50f41dca" width="100">

**Comment by Lar-Sen** [comment]
Mine matches exactly [LC_KGF_NT_V2.5](https://imgur.com/gallery/72CGuWt) : It has LC_KGF_NT_V2.5 marking, features AAE0 revision of nRF52, same capacitor/resistor count. Only visible difference is 32kHz oscillator which is labeled DA936C.

Yours has this ~~_strange_~~ QFAA-E1 (edit: E1 is perfectly legit rev.2) revision, seen in Amiloop specs. My assumption about capacitors missing may not be what's causing boot issues. Anyway I cannot check their values.
There must be some obscure differences between E0/E1 steepings which Amiloop designers could have exploited. It may look overkill, but in China they sometimes put anti-counterfeiting measures in very cheap products :)
Amiloop v4 may have a check routine for AAE1 version in FICR. Investigating...

APPROTECT Patch for AmiLoop (amiloop_update-230608.bin):
At offset 0xAFAB (ROM: 0x30FAB) put:
`10 F1 A6 0F 06 D0 6F F0  A5 01`
If you want to spare a useless softreset after unsetting APPROTECT, put this instead:
`10 F1 A6 0F 04 D0 6F F0 A5 01 02 48 00 F0 8F FB 10 BD 00 00 00`

**Comment by Lar-Sen** [comment]
Apparently no direct hardware revision checking.
When you have time, could you please dump the "good" amiibolink FICR/UICR region for me? I would like to compare factory and user fields between these 2 devices. I just need the 8kbytes starting at memory offset 0x10000000.
Thank you :)

**Comment by likaci** [comment]
@Lar-Sen No problem. Since my amiibolink has already erased and burned espruino, I don't have the original amiibolink firmware. This data is dumped using amiloop v3 mixed with amiibolink app as you described earlier.
[FICRUICR_v3loop_mix_link.bin.txt](https://github.com/bleem313/amiloop-files/files/14393251/FICRUICR_v3loop_mix_link.bin.txt)

**Comment by Lar-Sen** [comment]
Thank you
Nothing spectacular here, I don't see any difference in errata management, as mcu "magic" bytes are the same for both E0 and E1 (0x06). UICR, being erased in nrf52 recovery process, only shows same nReset GPIO mappings and ApProtect status.

There should be other differences in components values, making AmiiboLink crash if flashed with Amiloop fw. I don't see how  nrf52 version could be an issue.
